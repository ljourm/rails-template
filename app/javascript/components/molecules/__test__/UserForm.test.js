import { createLocalVue, mount } from "@vue/test-utils"
import Buefy from "buefy"
import Component from "../UserForm"

const email = "test@example.com"
const user = {
  uuid: "uuid",
  email: email,
  name: "name",
  roles: ["role_management"],
}
const roles = [
  { key: "role_management", name: "ロール管理" },
  { key: "user_management", name: "ユーザー管理" },
]

describe("Component", () => {
  const createWrapper = (user, roles) => {
    const localVue = createLocalVue()
    localVue.use(Buefy)

    return mount(Component, {
      localVue,
      propsData: { user, roles },
    })
  }

  it("test", async () => {
    const wrapper = createWrapper(user, roles)

    expect(wrapper.element).toMatchSnapshot()
  })

  describe("event", () => {
    const wrapper = createWrapper(user, roles)

    it("submit event", () => {
      wrapper.find("form").trigger("submit")
      expect(wrapper.emitted("submit")[0][0]).toEqual(user)
    })

    it("destroy event", () => {
      wrapper.find("button.destroy").trigger("click")
      expect(wrapper.emitted("destroy").length).toBe(1)
    })
  })

  describe("when empty user", () => {
    const wrapper = createWrapper({}, roles)

    it("should not be error", () => {
      expect(wrapper.vm.selectedRoles.length).toBe(0)
      expect(wrapper.vm.unselectedRoles.length).toBe(0)
    })
  })

  describe("when empty roles", () => {
    const wrapper = createWrapper(user, [])

    it("should not be error", () => {
      expect(wrapper.vm.selectedRoles.length).toBe(0)
      expect(wrapper.vm.unselectedRoles.length).toBe(0)
    })
  })

  describe("role chain", () => {
    const wrapper = createWrapper(user, roles)

    it("initial status", () => {
      const tagText = wrapper.find(".roles span.tag").text()
      expect(tagText).toBe("ロール管理")
      expect(tagText).not.toBe("ユーザー管理")

      expect(wrapper.vm.selectedRoles.length).toBe(1)
      expect(wrapper.vm.selectedRoles[0].key).toBe("role_management")
    })

    it("changed status and submit event", () => {
      wrapper.setData({ selectedRoles: [roles[1]] })

      wrapper.vm.$nextTick(() => {
        expect(wrapper.vm.selectedRoles.length).toBe(1)
        expect(wrapper.vm.selectedRoles[0].key).toBe("user_management")

        expect(wrapper.vm.unselectedRoles.length).toBe(1)
        expect(wrapper.vm.unselectedRoles[0].key).toBe("role_management")

        wrapper.find("form").trigger("submit")
        expect(wrapper.emitted("submit")[0][0].roles).toEqual([
          "user_management",
        ])
      })
    })
  })
})
