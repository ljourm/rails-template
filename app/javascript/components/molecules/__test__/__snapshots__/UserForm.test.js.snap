// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component test 1`] = `
<form>
  <div
    class="field"
  >
    <label
      class="label"
    >
      氏名
    </label>
    <div
      class="control"
    >
      <input
        aria-label="name"
        class="input"
        name="name"
        type="text"
      />
    </div>
  </div>
  <div
    class="field"
  >
    <label
      class="label"
    >
      EMail
    </label>
    <div
      class="control"
    >
      <input
        aria-label="email"
        class="input"
        name="email"
        type="text"
      />
    </div>
  </div>
  <div
    class="field"
  >
    <label
      class="label"
    >
      パスワード
    </label>
    <div
      class="control"
    >
      <input
        aria-label="password"
        class="input"
        name="password"
        type="password"
      />
    </div>
  </div>
  <div
    class="field"
  >
    <label
      class="label"
    >
      パスワード確認
    </label>
    <div
      class="control"
    >
      <input
        aria-label="password_confirmation"
        class="input"
        name="password_confirmation"
        type="password"
      />
    </div>
  </div>
  <div
    class="field"
  >
    <label
      class="label"
    >
      ロール
    </label>
    <div
      class="control"
    >
      <b-taginput
        allow-new="true"
        autocomplete="autocomplete"
        data=""
        field="name"
        icon="label"
        open-on-focus="true"
        readonly="true"
      />
    </div>
  </div>
  <div
    class="field is-grouped"
  >
    <div
      class="control"
    >
      <button
        class="button is-primary submit"
        type="submit"
      >
        送信
      </button>
    </div>
    <div
      class="control"
    >
      <button
        class="button destroy"
      >
        削除
      </button>
    </div>
  </div>
</form>
`;
